### ISSUE TITLE
How can I compile the sdk for Apple arm64 

### ISSUE THREAD

**Issue by AmyLee91** [Azure/azure-iot-sdk-c#2026]
My product need to support Apple M1.
I need to compile azure-iot-sdk-c for arm64.

Can you provide the guide of compiling the SDK for apple arm64?
Looking forward to your reply. Thank you.

**Comment by AmyLee91** [comment]
I try to compile the SDK on Apple M1.
OS:11.4
Xcode:12.2

my compile step:
`
- git clone -b LTS_01_2021_Ref01 https://github.com/Azure/azure-iot-sdk-c.git
- cd azure-iot-sdk-c
- git submodule update --init
- 
- cd azure-iot-sdk-c
- mkdir cmake
- cd cmake
- cmake -DOPENSSL_ROOT_DIR:PATH=~/openssl-OpenSSL_1_1_1i -Duse_openssl:bool=ON -DCMAKE_OSX_ARCHITECTURES="arm64" ..
- cmake --build .  

`

But the last step indicates some errors:

```
/Users/user/tmp/azure-iot-sdk/azure-iot-sdk-c/c-utility/adapters/x509_openssl.c:189:28: error: incomplete definition of type
      'struct evp_pkey_st'
            int evp_type = EVP_PKEY_id(evp_key);
                           ^           ~~~~~~~
/Users/user/tmp/azure-iot-sdk/azure-iot-sdk-c/c-utility/adapters/x509_openssl.c:18:45: note: expanded from macro 'EVP_PKEY_id'
        #define EVP_PKEY_id(evp_key) evp_key->type
                                     ~~~~~~~^
/Users/user/tmp/openssl/openssl-OpenSSL_1_1_1i/include/openssl/ossl_typ.h:93:16: note: forward declaration of 'struct evp_pkey_st'
typedef struct evp_pkey_st EVP_PKEY;
               ^
1 error generated.
make[2]: *** [c-utility/CMakeFiles/aziotsharedutil.dir/adapters/x509_openssl.c.o] Error 1
make[1]: *** [c-utility/CMakeFiles/aziotsharedutil.dir/all] Error 2
make: *** [all] Error 2

```

Could you tell me the cause of this error?

**Comment by danewalton** [comment]
Hi @AmyLee91. Thanks for the submission.

I should be able to get access to an M1 Mac shortly. I'll try and look into this

**Comment by danewalton** [comment]
Could you give your compilation options that you used for OpenSSL? I would like to try and repro as closely as possible.

**Comment by danewalton** [comment]
Okay this looks to be an issue with supporting OpenSSL version 1.1.1 ([linked issue here](#1996)). I get this error if I build on Intel Mac or M1 Mac.

The fix is here in `x509_openssl.c`:
Old
```c
#ifdef __APPLE__
    #ifndef EVP_PKEY_id
        #define EVP_PKEY_id(evp_key) evp_key->type
    #endif // EVP_PKEY_id
#endif // __APPLE__
```

New
```c
#ifdef __APPLE__
    #ifndef EVP_PKEY_id
        #include "crypto/evp.h"
        #define EVP_PKEY_id(evp_key) evp_key->type
    #endif // EVP_PKEY_id
#endif // __APPLE__
```

So the include `#include "crypto/evp.h"` needs to be added.

That along with the fix in #2029 should get you a successful build on Mac. 

I'm putting the fix here and not in a PR at the moment because this requires larger testing and support for OpenSSL 1.1.1 which could take more time. Hopefully this unblocks you for the moment.

**Comment by danewalton** [comment]
closing for now since the brunt of the issue is taken care of. Larger support of OpenSSL 1.1.1 is tracked in another issue.

**Comment by amarthere** [comment]
Hi Danewalton,

I am getting same issue but when i tried to add line "#include "crypto/evp.h",  i get file not found.
Please let me know where is the crypto/evp.h file exactly?

**Comment by danewalton** [comment]
This should in theory be in the newer version of OpenSSL for you.
